ecosystem/typescript/sdk_v2/src/bcs/deserializer.ts

```diff
@@ -0,0 +1,188 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+/* eslint-disable no-bitwise */
+import { MAX_U32_NUMBER } from "./consts";
+import { Bytes, Uint128, Uint16, Uint256, Uint32, Uint64, Uint8 } from "./types";
+
+export class Deserializer {
+  private buffer: ArrayBuffer;
+
+  private offset: number;
+
+  constructor(data: Bytes) {
+    // copies data to prevent outside mutation of buffer.
+    this.buffer = new ArrayBuffer(data.length);
+    new Uint8Array(this.buffer).set(data, 0);
+    this.offset = 0;
+  }
+
+  private read(length: number): ArrayBuffer {
+    if (this.offset + length > this.buffer.byteLength) {
+      throw new Error("Reached to the end of buffer");
+    }
+
+    const bytes = this.buffer.slice(this.offset, this.offset + length);
+    this.offset += length;
+    return bytes;
+  }
+
+  /**
+   * Deserializes a string. UTF8 string is supported. Reads the string's bytes length "l" first,
+   * and then reads "l" bytes of content. Decodes the byte array into a string.
+   *
+   * BCS layout for "string": string_length | string_content. string_length is the bytes length of
+   * the string that is uleb128 encoded. string_length is a u32 integer.
+   *
+   * @example
+   * ```ts
+   * const deserializer = new Deserializer(new Uint8Array([24, 0xc3, 0xa7, 0xc3, 0xa5, 0xe2, 0x88, 0x9e,
+   * 0xe2, 0x89, 0xa0, 0xc2, 0xa2, 0xc3, 0xb5, 0xc3, 0x9f, 0xe2, 0x88, 0x82, 0xc6, 0x92, 0xe2, 0x88, 0xab]));
+   * assert(deserializer.deserializeStr() === "çå∞≠¢õß∂ƒ∫");
```

Comment:
Lol, these bytes, maybe a more human readable example?

Reply:
Lol yea this was copied from the example in the original sdk idk why it was this string...

whenever we do do this we should use something like this:

```typescript
const deserializer = new Deserializer(new Uint8Array([6, 0x61, 0x62, 0x63, 0x31, 0x32, 0x33]));
assert(deserializer.deserializeStr() == "abc123");
```